marklogic-client-api/src/main/java/com/marklogic/client/datamovement/ProgressListener.java

```diff
@@ -0,0 +1,189 @@
+package com.marklogic.client.datamovement;
+
+import com.marklogic.client.datamovement.QueryBatch;
+import com.marklogic.client.datamovement.QueryBatchListener;
+import com.marklogic.client.datamovement.QueryBatcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+/**
+ * QueryBatchListener for receiving progress updates as QueryBatch instances are processed. Progress updates are
+ * communicated via instances of the nested ProgressUpdate interface.
+ */
+public class ProgressListener implements QueryBatchListener {
+
+	private static Logger logger = LoggerFactory.getLogger(ProgressListener.class);
+
+	private List<Consumer<ProgressUpdate>> consumers = new ArrayList<>();
+	private AtomicLong resultsSoFar = new AtomicLong(0);
+	private long startTime;
+	private long totalResults;
+
+	/**
+	 * Use this constructor for when the total number of results isn't known ahead of time.
+	 *
+	 * @param consumers
+	 */
+	public ProgressListener(Consumer<ProgressUpdate>... consumers) {
```

Comment:
Can we have the consumers registered via a new onProgressUpdate() method rather than taking the consumers in the constructor. That provides a little more flexibility when it comes to adding Consumers and it would be consistent with the other listeners. Something along the lines of 
```
public ProgressListener onProgressUpdate(Consumer<ProgressUpdate> consumer) 
```
and return ProgressListener so that we can chain them. Similarly for totalResults as well,
```
public ProgressListener withTotalResults(long totalResults) 
```

Reply:
Good point, I'll model it that way - I'll keep the constructor arg approach as well.

Reply:
Added chaining methods

Reply:
Thanks!